test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go

```diff
@@ -0,0 +1,459 @@
+	if err != nil {
+		log.Info("DockerMachinePool owner MachinePool is missing cluster label or cluster does not exist")
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		log.Info(fmt.Sprintf("Please associate this machine pool with a cluster using the label %s: <name of cluster>", clusterv1.ClusterLabelName))
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues("cluster", cluster.Name)
+
+	// Fetch the Docker Cluster.
+	dockerCluster := &infrav1.DockerCluster{}
+	dockerClusterName := client.ObjectKey{
+		Namespace: dockerMachinePool.Namespace,
+		Name:      cluster.Spec.InfrastructureRef.Name,
+	}
+	if err := r.Client.Get(ctx, dockerClusterName, dockerCluster); err != nil {
+		log.Info("DockerCluster is not available yet")
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues("docker-cluster", dockerCluster.Name)
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(dockerMachinePool, r)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Always attempt to Patch the DockerMachinePool object and status after each reconciliation.
+	defer func() {
+		if err := patchDockerMachinePool(ctx, patchHelper, dockerMachinePool); err != nil {
+			log.Error(err, "failed to patch DockerMachine")
+			if rerr == nil {
+				rerr = err
+			}
+		}
+	}()
+
+	// Add finalizer first if not exist to avoid the race condition between init and delete
+	if !controllerutil.ContainsFinalizer(dockerMachinePool, infrav1exp.MachinePoolFinalizer) {
+		controllerutil.AddFinalizer(dockerMachinePool, infrav1exp.MachinePoolFinalizer)
+		return ctrl.Result{}, nil
+	}
+
+	// Check if the infrastructure is ready, otherwise return and wait for the cluster object to be updated
+	if !cluster.Status.InfrastructureReady {
+		log.Info("Waiting for DockerCluster Controller to create cluster infrastructure")
+		return ctrl.Result{}, nil
+	}
+
+	// Handle deleted machines
+	if !dockerMachinePool.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, cluster, machinePool, dockerMachinePool, log)
+	}
+
+	// Handle non-deleted machines
+	return r.reconcileNormal(ctx, cluster, machinePool, dockerMachinePool, log)
+}
+
+// SetupWithManager will add watches for this controller
+func (r *DockerMachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	clusterToDockerMachinePools, err := util.ClusterToObjectsMapper(mgr.GetClient(), &infrav1exp.DockerMachinePoolList{}, mgr.GetScheme())
+	if err != nil {
+		return err
+	}
+
+	c, err := ctrl.NewControllerManagedBy(mgr).
+		For(&infrav1exp.DockerMachinePool{}).
+		WithOptions(options).
+		WithEventFilter(predicates.ResourceNotPaused(r.Log)).
+		Watches(
+			&source.Kind{Type: &clusterv1exp.MachinePool{}},
+			&handler.EnqueueRequestsFromMapFunc{
+				ToRequests: utilexp.MachinePoolToInfrastructureMapFunc(infrav1.GroupVersion.WithKind("DockerMachinePool")),
+			},
+		).
+		Watches(
+			&source.Kind{Type: &infrav1.DockerCluster{}},
```

Comment:
Question: Why are we watching for changes to DockerCluster?

Reply:
No good reason. Removed.